FirebaseFirestoreUI/FUISnapshotArrayDiff.h

```diff
@@ -162,8 +162,8 @@ FUIUnorderedPair *FUIUnorderedPairMake(id left, id right);
  * Creates a diff between two arrays. This operation is relatively expensive;
  * O(n x m) for arrays of length n and m.
  */
-- (instancetype)initWithInitialArray:(NSArray<ObjectType> *)initialArray
-                         resultArray:(NSArray<ObjectType> *)resultArray;
+- (instancetype)initWithInitialArray:(NSArray<FIRDocumentSnapshot *> *)initialArray
+                         resultArray:(NSArray<FIRDocumentSnapshot *> *)resultArray;
 
 /**
  * Creates a diff between two arrays, using the document changes array to speed up
```

Comment:
Unrelated to this PR, but I peeked at this method and I'm pretty confused about what's going on...  Why can't you use documentChanges basically as-is?  There are a bunch of cases handled in the implementation of this method that I can't make sense of (e.g. handling a FIRDocumentChangeTypeRemoved change where the removed document is in the result set for some reason).

If I'm just missing something, consider updating this comment to explain what's going on.

As a further aside (sorry), it's hard for me to reason about the 9 output arrays from this diff (deletedIndexes, deletedObjects, etc...).  It would be nice if the diff output was basically (or literally?) identical to our FIRQuerySnapshot.documentChanges list (just an array of FIRDocumentChange objects).  If there's some reason that structure doesn't work well for you, I'd be curious to hear about it. If it doesn't work for FirebaseUI, then maybe we need to rethink it.  FWIW- I'd be open to exposing the FIRDocumentChange constructor if it allowed interesting API reuse here.

Reply:
There's no particularly good reason the `documentChanges` info isn't better-utilized other than  when I wrote it I wasn't as familiar with the guarantees around `documentChanges`. As for the output arrays, it might make more sense in the context of [how they're used](https://github.com/firebase/FirebaseUI-iOS/blob/master/FirebaseFirestoreUI/FUIFirestoreCollectionViewDataSource.m#L101-L162).